File: frontend/src/lib/api/large-community-list.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface LargeCommunityListRule {
  rule_number: number;
  description?: string | null;
  action: string;  // permit|deny
  regex?: string | null;
}

export interface LargeCommunityList {
  name: string;
  description?: string | null;
  rules: LargeCommunityListRule[];
}

export interface LargeCommunityListConfig {
  large_community_lists: LargeCommunityList[];
  total: number;
}

export interface LargeCommunityListCapabilities {
  version: string;
  features: {
    basic: { supported: boolean; description: string };
    rules: { supported: boolean; description: string };
    actions: { supported: boolean; description: string };
  };
  version_notes: {
    identical_versions: string;
  };
  device_name?: string;
}

export interface LargeCommunityListBatchOperation {
  op: string;
  value?: string;
}

export interface LargeCommunityListBatchRequest {
  name: string;
  rule_number?: number;
  operations: LargeCommunityListBatchOperation[];
}

// ============================================================================
// API Service
// ============================================================================

class LargeCommunityListService {
  /**
   * Get capabilities based on VyOS version
   */
  async getCapabilities(): Promise<LargeCommunityListCapabilities> {
    return apiClient.get<LargeCommunityListCapabilities>("/vyos/large-community-list/capabilities");
  }

  /**
   * Get all large community lists configuration
   */
  async getConfig(refresh: boolean = false): Promise<LargeCommunityListConfig> {
    return apiClient.get<LargeCommunityListConfig>("/vyos/large-community-list/config", {
      refresh: refresh.toString(),
    });
  }

  /**
   * Refresh the cached configuration
   */
  async refreshConfig(): Promise<unknown> {
    return apiClient.post("/vyos/config/refresh");
  }

  /**
   * Execute batch operations
   */
  async batchConfigure(request: LargeCommunityListBatchRequest): Promise<unknown> {
    const result = await apiClient.post("/vyos/large-community-list/batch", request);
    await this.refreshConfig();
    return result;
  }

  /**
   * Delete an entire large community list
   */
  async deleteLargeCommunityList(name: string): Promise<unknown> {
    const operations: LargeCommunityListBatchOperation[] = [];
    operations.push({ op: "delete_large_community_list" });

    return this.batchConfigure({
      name,
      rule_number: 0, // Not used for delete_large_community_list
      operations,
    });
  }

  /**
   * Delete a specific rule from a large community list and renumber remaining rules to close gaps
   */
  async deleteRule(name: string, ruleNumber: number): Promise<unknown> {
    // Get current configuration
    const config = await this.getConfig(true);
    const communityList = config.large_community_lists.find(cl => cl.name === name);

    if (!communityList) {
      throw new Error(`Large community list ${name} not found`);
    }

    // Get remaining rules (excluding the one being deleted)
    const remainingRules = communityList.rules.filter(r => r.rule_number !== ruleNumber);

    if (remainingRules.length === 0) {
      // If no rules left, just delete the rule
      const operations: LargeCommunityListBatchOperation[] = [];
      operations.push({ op: "delete_rule" });

      return this.batchConfigure({
        name,
        rule_number: ruleNumber,
        operations,
      });
    }

    // Sort remaining rules by their current number
    const sortedRules = remainingRules.sort((a, b) => a.rule_number - b.rule_number);

    // Renumber sequentially starting from the lowest existing number
    // This closes gaps: if you have 105, 106, 107, 108 and delete 106,
    // result will be 105, 106 (was 107), 107 (was 108)
    const startingNumber = sortedRules[0].rule_number;
    const reorderRules = sortedRules.map((rule, index) => ({
      old_number: rule.rule_number,
      new_number: startingNumber + index,
      rule_data: rule,
    }));

    // Check if any rule actually needs renumbering
    const needsReorder = reorderRules.some(r => r.old_number !== r.new_number);

    if (!needsReorder) {
      // No gaps to close, just delete the rule directly
      const operations: LargeCommunityListBatchOperation[] = [];
      operations.push({ op: "delete_rule" });

      return this.batchConfigure({
        name,
        rule_number: ruleNumber,
        operations,
      });
    }

    // Use reorder endpoint which will delete all rules (including the one we want to delete)
    // and recreate them with new sequential numbers
    return this.reorderRules(name, reorderRules);
  }

  /**
   * Helper: Create a new large community list with a rule
   */
  async createLargeCommunityList(name: string, description: string | null, rule: Partial<LargeCommunityListRule>): Promise<unknown> {
    const operations: LargeCommunityListBatchOperation[] = [];

    // Create large community list
    operations.push({ op: "set_large_community_list" });

    // Add description
    if (description) {
      operations.push({ op: "set_large_community_list_description", value: description });
    }

    // Create rule
    operations.push({ op: "set_rule" });

    // Rule description
    if (rule.description) {
      operations.push({ op: "set_rule_description", value: rule.description });
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex) {
      operations.push({ op: "set_rule_regex", value: rule.regex });
    }

    return this.batchConfigure({
      name,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Helper: Update an existing large community list or rule
   */
  async updateLargeCommunityList(
    name: string,
    originalLargeCommunityList: LargeCommunityList,
    description: string | null,
    rule?: Partial<LargeCommunityListRule>,
    ruleNumber?: number
  ): Promise<unknown> {
    const operations: LargeCommunityListBatchOperation[] = [];

    // Update description
    if (description !== originalLargeCommunityList.description) {
      if (description) {
        operations.push({ op: "set_large_community_list_description", value: description });
      } else if (originalLargeCommunityList.description) {
        operations.push({ op: "delete_large_community_list_description" });
      }
    }

    // If updating a rule
    if (rule && ruleNumber !== undefined) {
      const originalRule = originalLargeCommunityList.rules.find(r => r.rule_number === ruleNumber);

      // Rule description
      if (rule.description !== originalRule?.description) {
        if (rule.description) {
          operations.push({ op: "set_rule_description", value: rule.description });
        } else if (originalRule?.description) {
          operations.push({ op: "delete_rule_description" });
        }
      }

      // Rule action
      if (rule.action && rule.action !== originalRule?.action) {
        operations.push({ op: "set_rule_action", value: rule.action });
      }

      // Rule regex
      if (rule.regex !== originalRule?.regex) {
        if (rule.regex) {
          operations.push({ op: "set_rule_regex", value: rule.regex });
        } else if (originalRule?.regex) {
          operations.push({ op: "delete_rule_regex" });
        }
      }
    }

    return this.batchConfigure({
      name,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Helper: Add a new rule to existing large community list
   */
  async addRule(name: string, rule: Partial<LargeCommunityListRule>): Promise<unknown> {
    const operations: LargeCommunityListBatchOperation[] = [];

    // Create rule
    operations.push({ op: "set_rule" });

    // Rule description
    if (rule.description) {
      operations.push({ op: "set_rule_description", value: rule.description });
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex) {
      operations.push({ op: "set_rule_regex", value: rule.regex });
    }

    return this.batchConfigure({
      name,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Update an existing rule
   */
  async updateRule(name: string, ruleNumber: number, rule: Partial<LargeCommunityListRule>): Promise<unknown> {
    const operations: LargeCommunityListBatchOperation[] = [];

    // Rule description
    if (rule.description !== undefined) {
      if (rule.description) {
        operations.push({ op: "set_rule_description", value: rule.description });
      } else {
        operations.push({ op: "delete_rule_description" });
      }
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex !== undefined) {
      if (rule.regex) {
        operations.push({ op: "set_rule_regex", value: rule.regex });
      } else {
        operations.push({ op: "delete_rule_regex" });
      }
    }

    return this.batchConfigure({
      name,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Reorder large community list rules
   */
  async reorderRules(largeCommunityListName: string, rules: Array<{ old_number: number; new_number: number; rule_data: LargeCommunityListRule }>): Promise<unknown> {
    const result = await apiClient.post("/vyos/large-community-list/reorder", {
      large_community_list_name: largeCommunityListName,
      rules: rules,
    });
    await this.refreshConfig();
    return result;
  }
}

export const largeCommunityListService = new LargeCommunityListService();

File: backend/routers/large_community_list/large_community_list.py
"""
Large Community List Router

API endpoints for managing VyOS large-community-list configuration.
Supports version-aware configuration for VyOS 1.4 and 1.5 (identical feature sets).
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import LargeCommunityListBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from utils.router_helpers import load_vyos_capabilities
import inspect

router = APIRouter(prefix="/vyos/large-community-list", tags=["large-community-list"])

# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Pydantic Models
# ============================================================================


class LargeCommunityListRule(BaseModel):
    """Large community list rule"""
    rule_number: int
    description: Optional[str] = None
    action: str = "permit"  # permit|deny
    regex: Optional[str] = None


class LargeCommunityList(BaseModel):
    """Complete large community list configuration"""
    name: str
    description: Optional[str] = None
    rules: List[LargeCommunityListRule] = []


class LargeCommunityListConfig(BaseModel):
    """Response containing all large community lists"""
    large_community_lists: List[LargeCommunityList] = []
    total: int = 0


class LargeCommunityListBatchOperation(BaseModel):
    """Single operation in a batch request"""
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")


class LargeCommunityListBatchRequest(BaseModel):
    """Model for batch configuration"""
    name: str = Field(..., description="Large community list name")
    rule_number: Optional[int] = Field(None, description="Rule number (optional)")
    operations: List[LargeCommunityListBatchOperation]


class ReorderRuleItem(BaseModel):
    """Single rule in a reorder request"""
    old_number: int = Field(..., description="Original rule number")
    new_number: int = Field(..., description="New rule number after reorder")
    rule_data: LargeCommunityListRule = Field(..., description="Complete rule configuration")


class ReorderLargeCommunityListRequest(BaseModel):
    """Model for reordering large community list rules"""
    large_community_list_name: str = Field(..., description="Large community list name")
    rules: List[ReorderRuleItem] = Field(..., description="List of rules with new order")


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations"""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_large_community_list_capabilities(request: Request):
    """
    Get feature capabilities based on device VyOS version.

    Returns feature flags indicating which operations are supported.
    Allows frontends to conditionally enable/disable features.
    """
    return await load_vyos_capabilities(
        request=request,
        feature_group=FeatureGroup.BGP_LARGE_COMMUNITY,
        builder_cls=LargeCommunityListBatchBuilder,
    )


# ============================================================================
# Endpoint 2: Config (Generalized Data)
# ============================================================================


@router.get("/config", response_model=LargeCommunityListConfig)
async def get_large_community_list_config(http_request: Request, refresh: bool = False):
    """
    Get all large-community-list configuration from VyOS in a generalized format.

    Args:
        refresh: If True, force refresh from VyOS. If False, use cache.

    Returns:
        Generalized configuration data optimized for frontend consumption
    """
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        # Navigate to policy -> large-community-list
        large_community_list_config = full_config.get("policy", {}).get("large-community-list", {})

        if not large_community_list_config:
            return LargeCommunityListConfig(large_community_lists=[], total=0)

        # Parse large community lists
        large_community_lists = []
        for name, cl_data in large_community_list_config.items():
            large_community_list = parse_large_community_list(name, cl_data)
            large_community_lists.append(large_community_list)

        return LargeCommunityListConfig(large_community_lists=large_community_lists, total=len(large_community_lists))

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


def parse_large_community_list(name: str, cl_data: dict) -> LargeCommunityList:
    """Parse large community list configuration from VyOS format."""
    description = cl_data.get("description")

    rules = []
    rules_raw = cl_data.get("rule", {})
    if rules_raw:
        for rule_num, rule_data in rules_raw.items():
            rule = parse_rule(int(rule_num), rule_data)
            rules.append(rule)

    return LargeCommunityList(
        name=name,
        description=description,
        rules=sorted(rules, key=lambda r: r.rule_number)
    )


def parse_rule(rule_number: int, rule_data: dict) -> LargeCommunityListRule:
    """Parse large community list rule from VyOS format."""
    description = rule_data.get("description")
    action = rule_data.get("action", "permit")
    regex = rule_data.get("regex")

    return LargeCommunityListRule(
        rule_number=rule_number,
        description=description,
        action=action,
        regex=regex
    )


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch")
async def large_community_list_batch_configure(http_request: Request, body: LargeCommunityListBatchRequest):
    """
    Execute a batch of configuration operations.

    Allows multiple changes in a single VyOS commit for efficiency.
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = LargeCommunityListBatchBuilder(version=version)

        # Process operations using inspect for dynamic method calls
        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = list(sig.parameters.keys())

            # Build arguments dynamically
            args = []

            # Add large community list name
            if "name" in params:
                args.append(body.name)

            # Add rule number if specified and method accepts it
            if body.rule_number and "rule" in params:
                args.append(str(body.rule_number))

            # Add operation value if provided
            if operation.value and len(params) > len(args):
                # Check remaining parameters
                remaining_params = params[len(args):]
                for param in remaining_params:
                    if param != "self":
                        args.append(operation.value)
                        break

            method(*args)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Configuration updated"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Reorder Endpoint
# ============================================================================


@router.post("/reorder")
async def reorder_large_community_list_rules(http_request: Request, body: ReorderLargeCommunityListRequest):
    """
    Reorder large community list rules by deleting and recreating them in a single commit.

    This endpoint efficiently reorders multiple rules by:
    1. Deleting all specified rules in reverse order
    2. Recreating them with new rule numbers
    All operations are executed in a single VyOS commit.

    Args:
        request: Reorder request containing large community list name and list of rules

    Returns:
        VyOSResponse with success/failure information
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = LargeCommunityListBatchBuilder(version=version)

        # Step 1: Delete all rules in reverse order
        rules_to_delete = sorted([r.old_number for r in body.rules], reverse=True)
        for old_number in rules_to_delete:
            builder.delete_rule(body.large_community_list_name, str(old_number))

        # Step 2: Recreate rules with new numbers
        for rule_item in body.rules:
            new_number = rule_item.new_number
            rule_data = rule_item.rule_data

            # Create the rule
            builder.set_rule(body.large_community_list_name, str(new_number))

            # Set action
            if rule_data.action:
                builder.set_rule_action(body.large_community_list_name, str(new_number), rule_data.action)

            # Set description
            if rule_data.description:
                builder.set_rule_description(body.large_community_list_name, str(new_number), rule_data.description)

            # Set regex
            if rule_data.regex:
                builder.set_rule_regex(body.large_community_list_name, str(new_number), rule_data.regex)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": f"Successfully reordered {len(body.rules)} rules in large community list {body.large_community_list_name}"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
